docs/planning/otel/specs/OTelWebSdk-Architecture.md

```diff
@@ -0,0 +1,2068 @@
+
+1. **Explicit Dependencies**: All components declare their dependencies upfront through factory function parameters
+2. **No Hidden Globals**: Zero reliance on global variables, window properties, or static state
+3. **Constructor Injection**: Dependencies provided at creation time, not accessed later
+4. **Interface-Based**: Components depend on interfaces, enabling easy testing and swapping
+5. **Lifecycle Clarity**: Parent components manage the lifecycle of their dependencies
+6. **Performance Optimization**: Closure-based implementation with injected dependencies for optimal runtime performance
+
+## Application Insights SDK Requirements
+
+Building on OpenTelemetry standards, the SDK inherits and extends the proven requirements from the Microsoft Application Insights JavaScript SDK to ensure enterprise-grade reliability and functionality:
+
+### 7. Dynamic Configuration Management
+- **Runtime Configuration Updates**: Support for configuration changes after SDK initialization without requiring restart
+- **Hot Reload Capability**: Ability to update connection strings, sampling rates, and feature flags dynamically
+- **Configuration Validation**: Real-time validation of configuration changes with rollback on failure
+- **Configuration Persistence**: Automatic persistence of valid configuration changes across application restarts
+- **Gradual Configuration Rollout**: Support for staged configuration deployment and A/B testing scenarios
+- **Reactive Configuration Pattern**: Configuration changes propagated through reactive event patterns rather than explicit APIs
+
+```typescript
+// Dynamic configuration implemented through reactive patterns
+const sdk = createOTelWebSdk(initialConfig);
+await sdk.initialize();
+
+// Configuration changes handled reactively via event callbacks
+sdk.onConfigChange((newConfig: IOTelWebSdkConfig) => {
+  console.log('Configuration updated:', newConfig);
+  // SDK automatically applies validated configuration changes
+});
+
+// Configuration updates can come from various sources:
+// - Remote configuration services
+// - Feature flag systems
+// - Environment variable changes
+// - Runtime API calls
+
+// Example: Remote configuration update triggers reactive pattern
+// External system pushes config change -> SDK validates -> onConfigChange fires -> Components adapt
+```
+
+### 8. Complete Unload and Cleanup Support
+- **Instance Unload**: Complete removal of SDK instances with all associated hooks, timers, and event listeners
+- **Memory Leak Prevention**: Comprehensive cleanup of closures, event handlers, and DOM modifications
+- **Hook Removal**: Automatic removal of all instrumentation hooks (XHR, fetch, unhandled errors, etc.)
+- **Resource Cleanup**: Proper disposal of timers, intervals, observers, and background processes
+- **Graceful Shutdown**: Coordinated shutdown ensuring in-flight telemetry is properly handled
+
+```typescript
+interface IUnloadManager {
+  unload(): Promise<IUnloadResult>;
```

Comment:
Most of these can be collapsed into a single function with passed / returned state instance and then using a callback.